tq/queue.go

```diff
@@ -0,0 +1,115 @@
+package tq
+
+import "sync"
+
+// Queue organizes and distributes work on OIDs against a set of available
+// workers (represented by a *WorkerQueue). Any failed items returned back by
+// that `*WorkerQueue` will be prioritized into the next batch.
+type Queue struct {
+	// incoming serves as the barrier between new calls to `Add()` and
+	// accumulating the next batch of items to be sent into the worker
+	// queue.
+	//
+	// incoming has a finite non-zero buffer equal to the size of a batch
+	// such that one extra batch of data can be buffered before applying
+	// back-pressure to callers.
+	incoming chan string
+	// workers maintains a handle on the *WorkerQueue instance used to
+	// distribute batched sets of OIDs
+	workers *WorkerQueue
+
+	// wg holds onto the run() goroutine and latches only after the
+	// goroutine has terminated, enabling safe shutdowns.
+	wg *sync.WaitGroup
+}
+
+// New instantiates a new `*tq.Queue` type with a given batch size "size",
+// number of workers "workers", and a function to do that work, "fn".
+//
+// Once returned, the `*Queue` will be active and able to accept new writes.
+func New(size, workers int, fn WorkerFn) *Queue {
+	q := &Queue{
+		incoming: make(chan string, size),
+		workers:  NewWorkerQueue(workers, fn),
+
+		wg: new(sync.WaitGroup),
+	}
+
+	q.wg.Add(1)
+	go q.run()
+
+	return q
+}
+
+// run accumulates new items into a batch for sending to the pool of available
+// workers, prioritizing retried objects over new objects.
+//
+// The function executes as follows:
+//
+//   1. Allocate a new batch of items for sending to the worker pool.
+//   2. Accept as many writes as we can before overfilling the batch from the
+//      `q.incoming` channel. If the channel closes, mark that we are closing
+//      and continue.
+//   3. Send the batch to the worker pool
+//   4. Empty the batch, begin re-filling it with items from the retry channel
+//      until the channel is closed.
+//   5. If closing, and the batch is empty, return.
+//   6. Otherwise, go to step 2.
+//
+// run runs in its own goroutine.
+func (q *Queue) run() {
+	defer q.wg.Done()
+
+	var closing bool
+
+	batch := make([]string, 0, cap(q.incoming))
```

Comment:
I think this should just hardcode 100. Pretty sure that's a hardcoded limit for batches. The batch size should not be determined by the size of the `q.incoming` buffer, since we may want to increase the buffer to estimate more objects in the mutable progress meter from #1747.